File: src/Harbor.Protocol/Packets.cs
using MessagePack;

namespace Harbor.Protocol;

// 주의: [MessagePackObject] 타입에서 한 줄 다중 필드 선언 금지 (같은 [Key] 공유됨).

// ----- DTO -----
[MessagePackObject]
public sealed class RoomDto
{
    [Key(0)] public long Id;
    [Key(1)] public string TemplateId = "";
    [Key(2)] public string Name = "";
    [Key(3)] public string[] Heightmap = Array.Empty<string>();
    [Key(4)] public short DoorX;
    [Key(5)] public short DoorY;
    [Key(6)] public string WallStyle = "";
    [Key(7)] public string FloorStyle = "";
    [Key(8)] public byte WallHeight = 3;      // 벽 높이(단위: 층). 클라 렌더용
    [Key(9)] public long OwnerId;            // 0 = 공용 방
    [Key(10)] public string OwnerNick = "";
    [Key(11)] public string Kind = "";       // "public" | "private"
    [Key(12)] public long UpgradePrice;      // 0 = 더 넓힐 수 없음
    [Key(13)] public string UpgradeName = ""; // 다음 단계 방 이름 (표시용)
    [Key(14)] public int Width;              // 타일 수 (표시용)
    [Key(15)] public int Height;
}

[MessagePackObject]
public sealed class ItemDto
{
    [Key(0)] public long Id;
    [Key(1)] public string FurniId = "";
    [Key(2)] public short X;
    [Key(3)] public short Y;
    [Key(4)] public float Z;
    [Key(5)] public byte Dir;
    [Key(6)] public string State = "";
    [Key(7)] public string? Extra;
    [Key(8)] public string Name = "";          // 표시 이름 (서버 정의 displayName)
    [Key(9)] public bool Solid;               // 통과 불가 여부 (클라 클릭 판정용)
    [Key(10)] public string Interaction = ""; // "fsm" | "seat" | "none"
    [Key(11)] public bool Wall;               // 벽걸이 가구 (dir 4 = 북쪽 벽, 2 = 서쪽 벽)
    [Key(12)] public string Author = "";      // 놓은 사람 닉 (포스트잇: 글쓴이. Extra = 본문)
    [Key(13)] public byte WallU;              // 벽 슬롯 가로 칸 (타일 모서리를 Walls.SlotCols 등분)
    [Key(14)] public byte WallV;              // 벽 슬롯 세로 칸 (0 = 바닥 쪽, 벽 높이 단위)
    [Key(15)] public sbyte Tilt;              // 기울기(도). 포스트잇 등 벽에 삐뚜름하게 붙이기
    [Key(16)] public bool Usable;             // 지금 상태에서 '사용'이 먹히는가 (자라는 중인 식물은 false)
}

[MessagePackObject]
public sealed class UserDto
{
    [Key(0)] public long Id;
    [Key(1)] public string Nick = "";
    [Key(2)] public string Figure = "";
    [Key(3)] public short X;
    [Key(4)] public short Y;
    [Key(5)] public float Z;
    [Key(6)] public byte Dir;
    [Key(7)] public string Action = "stand";
}

[MessagePackObject]
public sealed class RoomInfo
{
    [Key(0)] public long Id;
    [Key(1)] public string Name = "";
    [Key(2)] public string Kind = "";
    [Key(3)] public long OwnerId;
    [Key(4)] public string OwnerNick = "";
    [Key(5)] public int Users;
    [Key(6)] public int MaxUsers;
}

[MessagePackObject]
public sealed class CatalogEntry
{
    [Key(0)] public string FurniId = "";
    [Key(1)] public string Name = "";
    [Key(2)] public string Category = "";
    [Key(3)] public long Price;
    [Key(4)] public string Currency = "rupee";
    [Key(5)] public bool Wall;
    [Key(6)] public string Interaction = "";   // "postit" 이면 남의 방에도 붙일 수 있음
}

[MessagePackObject]
public sealed class InventoryEntry
{
    [Key(0)] public string FurniId = "";
    [Key(1)] public string Name = "";
    [Key(2)] public int Qty;
    [Key(3)] public bool Wall;
    [Key(4)] public string Interaction = "";
    [Key(5)] public long SellPrice;          // 0 = 팔 수 없음
    [Key(6)] public int SellBundleQty;       // 묶음 단위 (0 = 묶음 없음)
    [Key(7)] public long SellBundlePrice;
}

/// <summary>
/// 이사할 수 있는 집 한 채. 계열(family)마다 한 줄이고, 지금 단계와 같은 단계(없으면 그 계열의 마지막 단계)를 가리킨다.
/// </summary>
[MessagePackObject]
public sealed class HouseStyle
{
    [Key(0)] public string TemplateId = "";
    [Key(1)] public string Name = "";         // 그 단계 방 이름 ("넓은 모퉁이집")
    [Key(2)] public string Family = "";
    [Key(3)] public int Tier;
    [Key(4)] public int Tiles;                // 걸을 수 있는 칸 수 (얼마나 넓은지 비교용)
    [Key(5)] public long Price;               // 이사 비용
    [Key(6)] public bool Current;             // 지금 살고 있는 계열
}

[MessagePackObject] public sealed class TilePos { [Key(0)] public short X; [Key(1)] public short Y; }

// ----- C → S -----
[MessagePackObject] public sealed class C_Login { [Key(0)] public string Login = ""; [Key(1)] public string Token = ""; }
[MessagePackObject] public sealed class C_EnterRoom { [Key(0)] public long RoomId; }
[MessagePackObject] public sealed class C_Move { [Key(0)] public short X; [Key(1)] public short Y; }
[MessagePackObject] public sealed class C_Action { [Key(0)] public string Action = ""; }   // sit|stand|dance|wave|...
/// <summary>외모 변경. "hd-001-02.hr-002-05.ch-001-03.lg-001-01.sh-001-01.ha-001-04" 꼴 (파츠-모델-팔레트).</summary>
[MessagePackObject] public sealed class C_SetFigure { [Key(0)] public string Figure = ""; }
[MessagePackObject] public sealed class C_Chat { [Key(0)] public string Text = ""; [Key(1)] public byte Kind; }
[MessagePackObject] public sealed class C_Whisper { [Key(0)] public long TargetId; [Key(1)] public string Text = ""; }
[MessagePackObject] public sealed class C_PlaceItem { [Key(0)] public string FurniId = ""; [Key(1)] public short X; [Key(2)] public short Y; [Key(3)] public byte Dir; [Key(4)] public byte WallU; [Key(5)] public byte WallV; [Key(6)] public sbyte Tilt; }
[MessagePackObject] public sealed class C_PickItem { [Key(0)] public long ItemId; }
[MessagePackObject] public sealed class C_MoveItem { [Key(0)] public long ItemId; [Key(1)] public short X; [Key(2)] public short Y; [Key(3)] public byte Dir; }
[MessagePackObject] public sealed class C_UseItem { [Key(0)] public long ItemId; }
/// <summary>내가 붙인 포스트잇(ItemId)에 본문 쓰기. 서버가 S_ItemState(State="written", Extra=본문) 로 방송.</summary>
[MessagePackObject] public sealed class C_PostitWrite { [Key(0)] public long ItemId; [Key(1)] public string Body = ""; }
[MessagePackObject] public sealed class C_BuyCatalog { [Key(0)] public string FurniId = ""; [Key(1)] public int Qty = 1; }
/// <summary>가방의 수확물을 루피로. 묶음 단위가 있으면 묶음을 먼저 채운다(서버가 계산).</summary>
[MessagePackObject] public sealed class C_SellItem { [Key(0)] public string FurniId = ""; [Key(1)] public int Qty = 1; }
/// <summary>다른 계열의 집으로 이사. 방 모양이 달라지므로 **놓여 있던 가구는 전부 주인 가방으로 돌아온다.**</summary>
[MessagePackObject] public sealed class C_RemodelRoom { [Key(0)] public string TemplateId = ""; }

// ----- S → C -----
/// <summary>Nick 은 서버가 확정한 닉 — 입장권으로 들어오면 클라가 입력한 값과 다를 수 있으므로 이걸 따른다.</summary>
[MessagePackObject] public sealed class S_LoginResult { [Key(0)] public bool Ok; [Key(1)] public long UserId; [Key(2)] public string? Reason; [Key(3)] public long HomeRoomId; [Key(4)] public string Notice = ""; [Key(5)] public string Nick = ""; }
/// <summary>Rank 0 = 꽝. Prize 는 받은 루피(4등은 본전).</summary>
[MessagePackObject] public sealed class S_LotteryResult { [Key(0)] public int Rank; [Key(1)] public long Prize; [Key(2)] public long Price; [Key(3)] public string Message = ""; }
[MessagePackObject] public sealed class S_RoomSnapshot { [Key(0)] public RoomDto Room = new(); [Key(1)] public List<ItemDto> Items = new(); [Key(2)] public List<UserDto> Users = new(); }
[MessagePackObject] public sealed class S_RoomList { [Key(0)] public List<RoomInfo> Rooms = new(); }
[MessagePackObject] public sealed class S_HouseList { [Key(0)] public List<HouseStyle> Houses = new(); }
[MessagePackObject] public sealed class S_UserEnter { [Key(0)] public UserDto User = new(); }
[MessagePackObject] public sealed class S_UserLeave { [Key(0)] public long UserId; }
[MessagePackObject] public sealed class S_UserPath { [Key(0)] public long UserId; [Key(1)] public List<TilePos> Path = new(); }
[MessagePackObject] public sealed class S_UserAction { [Key(0)] public long UserId; [Key(1)] public string Action = ""; [Key(2)] public byte Dir; }
[MessagePackObject] public sealed class S_UserFigure { [Key(0)] public long UserId; [Key(1)] public string Figure = ""; }
[MessagePackObject] public sealed class S_ChatBubble { [Key(0)] public long UserId; [Key(1)] public string Text = ""; [Key(2)] public byte Kind; }
[MessagePackObject] public sealed class S_ItemAdd { [Key(0)] public ItemDto Item = new(); }
[MessagePackObject] public sealed class S_ItemRemove { [Key(0)] public long ItemId; }
[MessagePackObject] public sealed class S_ItemUpdate { [Key(0)] public ItemDto Item = new(); }
[MessagePackObject] public sealed class S_ItemState { [Key(0)] public long ItemId; [Key(1)] public string State = ""; [Key(2)] public string? Extra; [Key(3)] public bool Usable; }
/// <summary>Qty 는 변화량이 아니라 현재 보유 수량(절대값).</summary>
[MessagePackObject] public sealed class S_InventoryUpdate { [Key(0)] public string FurniId = ""; [Key(1)] public int Qty; [Key(2)] public string Name = ""; [Key(3)] public bool Wall; [Key(4)] public string Interaction = ""; [Key(5)] public long SellPrice; [Key(6)] public int SellBundleQty; [Key(7)] public long SellBundlePrice; }
/// <summary>서버가 보내는 일반 안내(토스트+채팅 로그). 오류가 아니다.</summary>
[MessagePackObject] public sealed class S_Notice { [Key(0)] public string Text = ""; }
[MessagePackObject] public sealed class S_Inventory { [Key(0)] public List<InventoryEntry> Items = new(); }
[MessagePackObject] public sealed class S_Catalog { [Key(0)] public List<CatalogEntry> Items = new(); }
[MessagePackObject] public sealed class S_WalletUpdate { [Key(0)] public long Rupee; [Key(1)] public long Cash; }
[MessagePackObject] public sealed class S_Error { [Key(0)] public ushort Code; [Key(1)] public string Message = ""; }

[MessagePackObject] public sealed class Empty { }

File: webapp/lib/protocol/packets.ts
// ⚠ **필드 이름은 전선에 실리지 않는다.**
//
// 서버의 `[MessagePackObject]` + `[Key(n)]` 은 객체를 **맵이 아니라 배열**로 직렬화한다.
// `C_Login { Login, Token }` 은 `{"Login":…}` 이 아니라 `[login, token]` 으로 나간다.
// 그래서 여기서는 **자리(index)** 로 읽고 쓴다. 이름으로 맞추려 들면 전부 undefined 가 되는데,
// 오류가 나지 않고 **조용히 빈 화면**이 된다.
//
// 따라서 `Packets.cs` 의 `[Key]` 번호는 계약이다. 중간에 필드를 끼워 넣으면 웹 클라가 통째로 어긋난다.
// 새 필드는 **반드시 맨 뒤 번호**로 붙인다.

import { Op } from "./opcode";

// ---------- 읽기 도우미 ----------
const arr = (v: unknown): unknown[] => (Array.isArray(v) ? v : []);
const num = (v: unknown, d = 0): number => (typeof v === "number" ? v : d);
const str = (v: unknown, d = ""): string => (typeof v === "string" ? v : d);
const bool = (v: unknown): boolean => v === true;
const strs = (v: unknown): string[] => arr(v).map((x) => str(x));

// ---------- DTO ----------
export interface RoomDto {
  id: number; templateId: string; name: string; heightmap: string[];
  doorX: number; doorY: number; wallStyle: string; floorStyle: string; wallHeight: number;
  ownerId: number; ownerNick: string; kind: string;
  upgradePrice: number; upgradeName: string;
  width: number; height: number;
  /** 한 칸 걷는 데 걸리는 ms. **이 값으로 보간해야** 서버 걸음과 어긋나지 않는다. */
  moveMs: number;
}

export function readRoom(v: unknown): RoomDto {
  const a = arr(v);
  return {
    id: num(a[0]), templateId: str(a[1]), name: str(a[2]), heightmap: strs(a[3]),
    doorX: num(a[4]), doorY: num(a[5]), wallStyle: str(a[6]), floorStyle: str(a[7]),
    wallHeight: num(a[8], 3), ownerId: num(a[9]), ownerNick: str(a[10]), kind: str(a[11]),
    upgradePrice: num(a[12]), upgradeName: str(a[13]),
    width: num(a[14]), height: num(a[15]), moveMs: num(a[16], 240),
  };
}

export interface ItemDto {
  id: number; furniId: string; x: number; y: number; z: number; dir: number;
  state: string; extra: string | null; name: string; solid: boolean;
  interaction: string; wall: boolean; author: string;
  wallU: number; wallV: number; tilt: number; usable: boolean;
}

export function readItem(v: unknown): ItemDto {
  const a = arr(v);
  return {
    id: num(a[0]), furniId: str(a[1]), x: num(a[2]), y: num(a[3]), z: num(a[4]), dir: num(a[5]),
    state: str(a[6]), extra: typeof a[7] === "string" ? a[7] : null, name: str(a[8]), solid: bool(a[9]),
    interaction: str(a[10]), wall: bool(a[11]), author: str(a[12]),
    wallU: num(a[13]), wallV: num(a[14]), tilt: num(a[15]), usable: bool(a[16]),
  };
}

export interface UserDto {
  id: number; nick: string; figure: string;
  x: number; y: number; z: number; dir: number; action: string;
  /** 인기도 — 내 화분에 남들이 꽂아 준 캣닢의 누적 개수. */
  fame: number;
}

export function readUser(v: unknown): UserDto {
  const a = arr(v);
  return {
    id: num(a[0]), nick: str(a[1]), figure: str(a[2]),
    x: num(a[3]), y: num(a[4]), z: num(a[5]), dir: num(a[6]), action: str(a[7], "stand"),
    fame: num(a[8]),
  };
}

export interface TilePos { x: number; y: number }
const readTile = (v: unknown): TilePos => { const a = arr(v); return { x: num(a[0]), y: num(a[1]) }; };

// ---------- S → C ----------
export interface LoginResult {
  ok: boolean; userId: number; reason: string | null; homeRoomId: number; notice: string;
  /** 서버가 확정한 닉. 입장권으로 들어오면 입력값과 다를 수 있으니 **이것을 따른다.** */
  nick: string;
}

export const readLoginResult = (v: unknown): LoginResult => {
  const a = arr(v);
  return {
    ok: bool(a[0]), userId: num(a[1]), reason: typeof a[2] === "string" ? a[2] : null,
    homeRoomId: num(a[3]), notice: str(a[4]), nick: str(a[5]),
  };
};

export interface RoomSnapshot { room: RoomDto; items: ItemDto[]; users: UserDto[] }

export const readRoomSnapshot = (v: unknown): RoomSnapshot => {
  const a = arr(v);
  return { room: readRoom(a[0]), items: arr(a[1]).map(readItem), users: arr(a[2]).map(readUser) };
};

export const readError = (v: unknown) => { const a = arr(v); return { code: num(a[0]), message: str(a[1]) }; };
export const readNotice = (v: unknown) => ({ text: str(arr(v)[0]) });
export const readUserEnter = (v: unknown) => ({ user: readUser(arr(v)[0]) });
export const readUserLeave = (v: unknown) => ({ userId: num(arr(v)[0]) });
export const readUserPath = (v: unknown) => {
  const a = arr(v);
  return { userId: num(a[0]), path: arr(a[1]).map(readTile) };
};
export const readUserAction = (v: unknown) => {
  const a = arr(v);
  return { userId: num(a[0]), action: str(a[1]), dir: num(a[2]) };
};
export const readChat = (v: unknown) => {
  const a = arr(v);
  return { userId: num(a[0]), text: str(a[1]), kind: num(a[2]) };
};
export const readWallet = (v: unknown) => { const a = arr(v); return { rupee: num(a[0]), cash: num(a[1]) }; };

// ---------- 방 목록 ----------
export interface RoomInfo {
  id: number;
  name: string;
  /** "public" = 누구나 오는 광장·댄스홀. "private" = 개인 집. */
  kind: string;
  ownerId: number;
  ownerNick: string;
  users: number;
  maxUsers: number;
}

export const readRoomInfo = (v: unknown): RoomInfo => {
  const a = arr(v);
  return {
    id: num(a[0]), name: str(a[1]), kind: str(a[2]),
    ownerId: num(a[3]), ownerNick: str(a[4]),
    users: num(a[5]), maxUsers: num(a[6], 25),
  };
};

export const readRoomList = (v: unknown): RoomInfo[] => arr(arr(v)[0]).map(readRoomInfo);

export const roomListPacket = () => ({ op: Op.C_RoomList, body: [] });

// ---------- 가방 ----------
export interface InvEntry {
  furniId: string;
  name: string;
  qty: number;
  wall: boolean;
  interaction: string;
  /** 0 이면 팔 수 없다(수확물만 팔린다). */
  sellPrice: number;
  /** 묶음 단위. 0 이면 묶음 없음. 묶음이 낱개×개수보다 비싸야 "모아서 파는" 의미가 있다. */
  sellBundleQty: number;
  sellBundlePrice: number;
}

/**
 * ⚠ **`S_Inventory` 의 항목과 `S_InventoryUpdate` 는 자리 순서가 다르다.**
 *   InventoryEntry    : [furniId, **name**, **qty**, wall, …]
 *   S_InventoryUpdate : [furniId, **qty**,  **name**, wall, …]
 * 1번과 2번이 뒤바뀌어 있다. 같은 읽기 함수를 쓰면 이름 자리에 숫자가 들어간다.
 */
const readInvEntryTail = (a: unknown[], furniId: string, name: string, qty: number): InvEntry => ({
  furniId, name, qty,
  wall: bool(a[3]), interaction: str(a[4]),
  sellPrice: num(a[5]), sellBundleQty: num(a[6]), sellBundlePrice: num(a[7]),
});

export const readInvEntry = (v: unknown): InvEntry => {
  const a = arr(v);
  return readInvEntryTail(a, str(a[0]), str(a[1]), num(a[2]));
};

/** S_InventoryUpdate — **Qty 는 변화량이 아니라 현재 보유 수량(절대값)**. 0 이면 목록에서 뺀다. */
export const readInvUpdate = (v: unknown): InvEntry => {
  const a = arr(v);
  return readInvEntryTail(a, str(a[0]), str(a[2]), num(a[1]));
};

export const readInventory = (v: unknown): InvEntry[] => arr(arr(v)[0]).map(readInvEntry);

// ---------- C → S (가방) ----------
export const inventoryPacket = () => ({ op: Op.C_Inventory, body: [] });
export const sellPacket = (furniId: string, qty = 1) => ({ op: Op.C_SellItem, body: [furniId, qty] });
export const placeItemPacket = (furniId: string, x: number, y: number, dir = 2, wallU = 0, wallV = 0, tilt = 0) =>
  ({ op: Op.C_PlaceItem, body: [furniId, x, y, dir, wallU, wallV, tilt] });

// ---------- C → S ----------
// 배열의 **자리**가 [Key] 번호다. 빠뜨리면 그 자리가 null 로 가서 서버가 기본값을 본다.
export const loginPacket = (login: string, token: string) => ({ op: Op.C_Login, body: [login, token] });
export const enterRoomPacket = (roomId: number) => ({ op: Op.C_EnterRoom, body: [roomId] });
export const movePacket = (x: number, y: number) => ({ op: Op.C_Move, body: [x, y] });
export const actionPacket = (action: string) => ({ op: Op.C_Action, body: [action] });
export const chatPacket = (text: string, kind = 0) => ({ op: Op.C_Chat, body: [text, kind] });
export const pingPacket = () => ({ op: Op.C_Ping, body: [] });
